File: leetcode/problems/322/322.cc
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp;
        map<int, int> coin;
        int mc = 0;
        dp.resize(200);
        for(int i = 0; i < coins.size(); i++){
            coin[coins[i]] = 1;
            mc = max(mc, coins[i]);
        }
        dp.resize(mc);
        dp[1] = INT_MAX;
        int a = 0;
        for(int j = 1; j <= amount; j++){
            for(int i = 0; i < coins.size(); i++){
                // Exact coin
                if(coins[i] == j){
                    dp[j] = 1;
                } else if (coins[i] < j){
                    cout << j << "da " << endl;
                // Make the change
                    a = 1 + dp[j - coins[i]];
                    if(a < dp[j]){
                        dp[j] = a;
                    }
                }
            }
            cout << j << " " << dp[j] << endl;
        }
        if(dp[amount] == 0){
            return -1;
        }
        return dp[amount];
    }
};

File: leetcode/problems/322/322.py
class Solution:
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [9999 for i in range(0, amount+1)]
        if amount == 0:
            return 0
        coins.sort()
        for i in range(0, amount+1):
            for j in range(0, len(coins)):
                # Each coin valid
                if coins[j] > i:
                    break
                elif coins[j] == i:
                    dp[i] = 1
                    break
                else:
                    dp[i] = min(dp[i - coins[j]] + 1, dp[i])
        if dp[amount] == 9999:
            return -1
        else:
            return dp[amount]